core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java

```diff
@@ -401,8 +402,18 @@ private void handleRequest(ChannelHandlerContext ctx, DecodedHttpRequest req) th
             }
 
             serviceResponse = serviceResponse.recover(cause -> {
-                // Recover a failed response with the default exception handler.
-                return exceptionHandler.convert(reqCtx, cause);
+                // Recover a failed response with the exception handler.
+                final HttpResponse converted = exceptionHandler.convert(reqCtx, cause);
+                assert converted != null;
+                if (converted instanceof AbortedHttpResponse) {
+                    return converted;
+                }
+                // Wraps the converted HttpResponse using HttpResponse.ofFailure(...) so that
+                // - We can log the responseCause.
+                // - The log is complete even though deferred values of such as
+                //   responseContent and responseContentPreview are not set.
+                // See https://github.com/line/armeria/issues/3719.
+                return HttpResponse.ofFailure(HttpResponseException.of(converted, cause));
```

Comment:
Do we need to log the exception if a cause is handled by a user? I think the following example is enough.
```java
HttpResponse.ofFailure(HttpResponseException.of(converted));
```

I think we can guide users:
- Use normal responses such as `HttpResonse.of(HttpStatus.BAD_REQUEST)` when a cause is fully recovered by an exception handler and will not log an exception.
- Use failed responses such as `HttpResonse.ofFailure(HttpStatus.BAD_REQUEST, cause)` or `HttpResonse.ofFailure(HttpResponse.of(...), cause)` to recover and log the exception.
  - The above factories are shortcut for `HttpResonse.ofFailure(HttpStatusException.of(HttpStatus.BAD_REQUEST), cause)` and `HttpResonse.ofFailure(HttpResponse.of(...), cause)`.

Reply:
Well, I thought exactly the opposite.
I thought users want to set the `log.responseCause()` in most cases.
So users just return the normal response in the `ExceptionHandler`. (They do not have to use `HttpResponse.ofFailure(HttpResponseException.of(..., cause))` which is a bit verbose)
Besides, we provide a way to modify the response cause log using `responseCauseSanitizer` so users could use that too if they do not want to log it.

Reply:
Could you explain why we can't just 'un-defer' the deferred properties when `ExceptionHandler` handles an exception?

Reply:
As we discussed, the `cause` is set to `log.responseCause()` by default. 😉 
If users wants not to set the `cause`, they need to:
- Return `HttpResponse` using `HttpResponseException` in the `ExceptionHandler`
  (e.g.) `HttpResponse.ofFailure(HttpResponseException.of(custom));`
- Or sanitize the cause using `LoggingService`